commit 715d43fedd811fd6869c3aef69f01dede4379403
Author: Andreas Jaeger <aj@suse.de>
Date:   Fri Aug 17 07:18:34 2007 +0000

    Make _nl_category_name_idxs extern.

diff --git a/locale/localeinfo.h b/locale/localeinfo.h
index 5c1f27f7e1..5e3e99ca68 100644
--- a/locale/localeinfo.h
+++ b/locale/localeinfo.h
@@ -1,5 +1,5 @@
 /* Declarations for internal libc locale interfaces
-   Copyright (C) 1995-2003, 2005, 2006 Free Software Foundation, Inc.
+   Copyright (C) 1995-2003, 2005, 2006, 2007 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
 
    The GNU C Library is free software; you can redistribute it and/or
@@ -186,7 +186,7 @@ extern const union catnamestr_t
   };
   char str[0];
 } _nl_category_names attribute_hidden;
-const uint8_t _nl_category_name_idxs[__LC_LAST] attribute_hidden;
+extern const uint8_t _nl_category_name_idxs[__LC_LAST] attribute_hidden;
 extern const uint8_t _nl_category_name_sizes[__LC_LAST] attribute_hidden;
 
 /* Name of the standard locales.  */
